GPII-3707: Move keyring from global to infra_region

mrtyler opened the pull request:
**Rendered docs: https://github.com/mrtyler/gpii-infra/blob/gpii-3707-regionalism/gcp/README.md**

Requires https://github.com/gpii-ops/exekube/pull/49, which can be safely merged on its own.
- [x] Merged and deployed

I'm finalizing documentation and the rollout plan but I think I'm done with the implementation, hence no WIP label.

### Changes
I tried to make targeted commits but there was a hard-to-remove detour where I thought I needed to change the Keyring name based on Region. Anyway, the main themes, ordered roughly by batches of commits:

1. Small cleanups, including removing Stepan's hack for GPII-3568. I think Stepan is removing that in one of his in-flight branches, but it was causing me trouble so I removed it here as well.
2. Refactor secrets.rb and its call sites in rake land. Use Keyrings in `infra_region` instead of `global`. Add some basic unit tests. I'm not a huge fan of classes but:
   * This is a classic use case for a class: data bundled with methods for working with that data
   * Dealing with environment variables in unit tests is kind of annoying (see vars_spec.rb).
   * I think instance variables (set via constructer parameters) are a more natural fit than environment variables, and they declutter the rake file a bit compared with passing things like `project_id` in every time.
3. Add some helpers to facilitate migration from the global keyring, and for moving between regions.

### Alternative approach
My main goal was to enable us to run GPII Cloud environments in any region. I have been able to do that in my tests, but the current approach has some downsides:
* It requires some extra `terraform import` steps because KMS Keyrings are indestructible and Keys have an unusual lifecycle
   * I think moving from region to region will be rare in practice
   * I've added some rudimentary automation for this, `rake import_keyring`
* Each environment has a single tfstate bucket in Multi-Region `US`, no matter in which Region the environment really runs

I have considered, but not yet tested, an alternative approach with a tfstate bucket in each region where we might wish to run an environment. This avoids the `terraform import` steps because tfstate in us-east1 can always point to the keyring in us-east1. I haven't pursued this avenue because:
* We need to migrate off of `global` Keyrings regardless, so something like this PR and its rollout plan will be necessary
* tfstate buckets are managed by common, so we'd need to create them ahead of time for each environment X region we want to have available
* Multiple tfstate buckets would allow (in fact, require) `apply_infra` resources (e.g. VPCs, public IPs) in each Region. This is a small waste and a small amount of extra clutter
* Multiple tfstate buckets would allow users to accidentally run two environments at once (e.g. one in us-east1, one in us-central1)
* DNS is still a singleton. The default behavior with multiple tfstate buckets would be "last one wins", which is not so bad but nonetheless adds a potential source of confusion.

### Exported logs
We have a decision to make about existing exported logs for stg and prd:
1. Delete them - how much value do they have?
1. Leave them alone - it is theoretically possible to decrypt and read the logs later as long as we are careful not to destroy the global Key used to write them
1. Re-encrypt them in place - this is the safest, but most expensive option
   * I think we can use a modified version of `rotate_tfstate_key` to accomplish this, but I haven't tested it yet.
- [x] Decided to keep prd logs

### Rollback plan
The basic idea is to undo the rollout by re-importing the `global` Keyring and Keys and re-encrypting data using the Old Keys. The steps here are for pointing the environment back to the Old Keyring. Other steps may be needed as well, depending on when the rollout was abandoned.

1. Revert this PR (i.e. go back to Old version of code)
   * The associated Exekube PR and resulting image may remain
1. Update gpii-infra to Old version
1. `cd gpii-infra/gcp/live/dev`
1. `rake sh`
1. `unset GOOGLE_ENCRYPTION_KEY`
1. `cd /project/live/$ENV/secret-mgmt`
1. Replace the New Keyring and Keys with the Old Keyring and Keys:
```
terragrunt state rm module.gcp-secret-mgmt.google_kms_key_ring.key_ring
terragrunt import module.gcp-secret-mgmt.google_kms_key_ring.key_ring projects/$TF_VAR_project_id/locations/global/keyRings/keyring
terragrunt state rm module.gcp-secret-mgmt.google_kms_crypto_key.encryption_keys[0]
terragrunt import module.gcp-secret-mgmt.google_kms_crypto_key.encryption_keys[0] projects/$TF_VAR_project_id/locations/global/keyRings/keyring/cryptoKeys/default
terragrunt state rm module.gcp-secret-mgmt.google_kms_crypto_key.encryption_keys[1]
terragrunt import module.gcp-secret-mgmt.google_kms_crypto_key.encryption_keys[1] projects/$TF_VAR_project_id/locations/global/keyRings/keyring/cryptoKeys/gcp-stackdriver-export
```

### Rollout plan
- [ ] Complete migration steps (below) for dev-doe (skipping some steps since dev-doe is an ephemeral cluster), which will be deployed into a new region (us-east4) as a test case
- [ ] Notify cloud-developers of required actions (see below)
- [ ] Merge this PR, which will eventually run CI for dev-doe.
- [ ] Verify everything looks correct. In particular, look out for changes to Keyrings and changes to the tfstate bucket as part of common-stg
- [ ] Complete migration steps (below) for stg
- [ ] Promote common-prd
- [ ] Verify everything looks correct in stg and in plan-all for prd
- [ ] Complete migration steps (below) for prd
- [ ] Promote prd
- [ ] Re-encrypt prd's exported logs (see below)

#### Migration steps

"Old" refers to the tip of master / the latest deployed infrastructure from before this PR (i.e. with Keyrings and Keys in the "global" Region). "New" refers to code / infrastructure after this PR is merged.

1. Update to latest version of gpii-infra
1. `cd gpii-infra/gcp/live/dev`
1. `rake update_exekube`
1. Replace the Old Keyring and Keys with the New Keyring and Keys:
   * `TF_VAR_decrypt_with_key_from_region=global rake import_keyring`
1. Verify that `terragrunt plan` doesn't want to make any changes
   * `TF_VAR_decrypt_with_key_from_region=global rake sh` and then inside shell:
   * `unset GOOGLE_ENCRYPTION_KEY`
   * `cd /project/live/$ENV/secret-mgmt`
   * `terragrunt plan`
   * Exit shell
1. `TF_VAR_decrypt_with_key_from_region=global rake rotate_secrets_key`
1. `rake rotate_secrets_key"[gcp-stackdriver-export]"`
1. `rake` (perhaps via CI)
   * Should be mostly a no-op.
1. Fix encryption for exported logs bucket (see below)
1. Verify logs exported both before and after the migration are readable
   * [Go to bucket](https://console.cloud.google.com/storage/browser/gpii-gcp-prd-exported-logs?project=gpii-gcp-prd&authuser=1&organizationId=247149361674)
   * Navigate to some Old logs. Verify that you can view them (no KMS error)
   * Navigate to some New logs -- you may have to wait up to an hour for new logs to be exported. Verify that you can view them (no KMS error)
1. Disable (do *NOT* destroy!) Keys in `global` Keyring

#### Email to cloud-developers
```
Subject: New regions available for dev envs (action required)


Hi cloud developers,

Good news: It is now possible to create a dev environment in any Google Cloud region![1][2] Instructions are here:
https://github.com/gpii-ops/gpii-infra/blob/master/gcp/README.md#i-want-to-spin-up-my-dev-environment-in-a-different-region

Bad news: You must take action -- even if you are happy to stay in your current region -- to keep your cluster up-to-date. (Please do this migration separately, before trying out a new region. :))

If you are able to destroy your environment, the process is pretty standard:
* `cd gpii-infra/gcp/live/dev`
* `rake destroy`
* Update gpii-infra to latest master
* `rake update_exekube`
* `rake destroy_hard`[3]
* `rake`

If you are not able to destroy your environment, you can safely migrate an existing environment. It's a little more complex, so you may want to consult Ops before pursuing this route:
* See "Rollout plan: Migration steps" in https://github.com/gpii-ops/gpii-infra/pull/329.

Please ping us in #ops if you have questions or any trouble!

Thanks,
the ops team


[1] I haven't tested anything outside of the US. Who will spin up the first GPII Cloud in Europe??

[2] Some infrastructure (the GCS Bucket where Terraform keeps its state) remains in the `US` multi-region.

[3] `destroy_hard` is a new helper command for destroying even more stuff related to your environment, like Terraform state and secrets.
```

#### Re-encrypt exported logs
TBD

stepanstipl commented:
Thanks for thorough description 👍, it helps ;). I'm commenting on the text of PR only so far:

> an alternative approach with a tfstate bucket in each region where we might wish to run an environment

What about creating the bucket in only the desired region (this would be parametrized in the common code) and then creating the cluster and all the resources in the corresponding region? (info about the region can be retrieved for example by using TF `remote_state`) That would avoid having unused resources and solve most/all the issues mentioned.

In general I would definitely prefer to have tfstate bucket in the same region as the rest of resources.

> We have a decision to make about existing exported logs for stg and prd:

As discussed, I'm ok to loose log in non-prod env. For production I would try to preserve those - i think it should be achieveavble using `gsutil` - unless the old key material is destroyed. gsutil shoul be able to read those logs and by running `cp` save & reencrypt them with the new key/keyring (`gsutil cp` between 2 GCS buckets should work directly without actually pulling/pushing data via the node running the command).

mrtyler commented:
Stepan, I'm afraid I don't understand your suggestion:

> What about creating the bucket in only the desired region (this would be parametrized in the common code) 

There is no single "desired region". The goal of this feature is to make it possible to run infrastructure in any region. So AFAICT we must either "migrate" back and forth, updating tfstate each time, or we must have dedicated tfstate in each region.

> and then creating the cluster and all the resources in the corresponding region? (info about the region can be retrieved for example by using TF remote_state) 

We do (with this PR) create the cluster and all resources in the specified region (i.e. the value of `TF_VAR_infra_region`. The only thing that is hardcoded to us-central1 is the tfstate bucket itself.

> That would avoid having unused resources and solve most/all the issues mentioned.

I don't see how.

mrtyler commented:
Re: exported logs:

Due to [an incident](https://pad.gpii.net/mypads/?/mypads/group/gpii-infrastructure-standups-lix4njm/pad/view/postmortems-but4nne#L255), we have already destroyed the exported stg logs.

For prd logs, we agreed to rewrite them and keep them. I am figuring out how best to do this.

stepanstipl commented:
> There is no single "desired region". The goal of this feature is to make it possible to run infrastructure in any region. So AFAICT we must either "migrate" back and forth, updating tfstate each time, or we must have dedicated tfstate in each region.

So where I was going with this was that the tfstate region, and therefore the region for the cluster, would be specified in common code. That way you can create the bucket in the region where the cluster will land.

So you would still be able to have cluster in any region, but you would want to change the region, you would have to rerun common.

mrtyler commented:
Passing the desired region to all the places that need it is not the problem. The problem is keeping tfstate files in sync with cloud resources. I also don't like forcing developers to ask for help if they want to change regions.

I'd like to move forward with the current design. No matter what we do with the tfstate bucket, we will do it in a separate PR as I think this PR already has enough moving parts.

stepanstipl commented on gcp/modules/gcp-secret-mgmt/main.tf:
```diff
@@ -20,6 +20,7 @@ module "gcp-secret-mgmt" {
   encryption_keys    = "${var.encryption_keys}"
   storage_location   = "${var.infra_region}"
   keyring_name       = "${var.keyring_name}"
```
Would it make sense to include the region in the name of the keyring? `"keyring-${var.infra_region}" maybe? This should eventually allow multiple keyrings in different regions to co-exist.

mrtyler replied:
It turns out that Keyrings in different regions need not have unique names. My dev environment has four Keyrings in different regions, all named "keyring".

stepanstipl replied:
ok then, I did not know that :), thanks for info

amatas commented:
LGTM, for now. 

I can not see very clear the advantages of having a bucket per region, and I can see some pain every time that we need to move from one to other. I would leave this PR as is, and see how we go. New approaches can be implemented in the future if we need them.

mrtyler commented:
> I would still consider unique names for keyrings (I think it'll make our lives easier).

This would avoid the need to import existing Keyrings when starting fresh. But Keyrings are indestructible, so we would accrue many discarded Keyrings over time. I think Sergey and I discussed this option and abandoned it for that reason.

mrtyler commented:
Ok, I've done more testing, fixed a couple bugs, and updated the docs and rollout plan. My plan is to deploy this tomorrow morning unless there are objections.

The last bit I need to work out is re-encrypting the exported logs. I don't think it will be a big deal, but I keep forgetting to test it before deleting my environment and all the exported logs in it :). This step is not in the critical path, so it can be delayed a bit while I get the rest of the changes out.

amatas commented:
Latest changes LGTM
